base/src/main/java/com/mindera/skeletoid/widgets/tooltip/PopupTooltip.kt

```diff
@@ -0,0 +1,466 @@
+
+        /**
+         * The LOGGING TAG. Epic.
+         */
+        const val TAG = "PopupTooltip"
+
+        /**
+         * Saves a reference to our currently displayed tooltip. This is here so we can control when
+         * we dismiss/show it.
+         */
+        private var tooltip: PopupWindow? = null
+
+        /**
+         * The estimated height of our popup in dips (density-independent pixels).
+         * This assumes the popup will have two lines of text with the text sizes
+         * specified in [layout].
+         */
+        const val ESTIMATED_HEIGHT_DIPS = 46
+
+        /**
+         * The default width value of the tooltip if none is specified, in dips (density-independent pixels).
+         */
+        const val TOOLTIP_WIDTH = 250
+
+        /**
+         * The default width value of the tooltip arrow if none is specified, in dips (density-independent pixels).
+         */
+        const val ARROW_WIDTH = 10
+
+        /**
+         * The default height value of the tooltip arrow if none is specified, in dips (density-independent pixels).
+         */
+        const val ARROW_HEIGHT = 10
+
+        /**
+         * The default elevation of the tooltip, in dips (density-independent pixels).
+         */
+        const val DEFAULT_ELEVATION = 5
+
+        /**
+         * At the moment this is hardcoded here. In the future the idea is for this
+         * to be changeable.
+         */
+        const val GRAVITY = Gravity.NO_GRAVITY
+
+        /**
+         * Sets the tooltip as null after showDuration amount of time. This is only used
+         * when we set the tooltip as definite.
+         */
+        private val handler = Handler()
+
+        /**
+         * Clears the current reference to the tooltip and dismisses if it is present.
+         */
+        fun clearTooltipIfNotNull() {
+            tooltip?.apply {
+                if (isShowing) dismiss()
+                tooltip = null
+            }
+        }
+
+        /**
+         * Given a value in dips, gets the amount of pixels that equates to in the given context.
+         */
+        fun getValueInPixels(valueInDips: Int, context: Context): Float {
+            return (valueInDips * context.resources.displayMetrics.density)
+        }
+    }
+
+    /**
+     * The vertical offset from the anchorview to the tooltip.
+     *
+     * @param margin - the margin in dips.
+     */
+    fun verticalMargin(margin: Int) = apply { this.verticalMargin = margin }
+
+    /**
+     * The horizontal offset from the edges of the screen.
+     *
+     * @param margin - the margin in dips.
+     */
+    fun horizontalMargin(margin: Int) = apply { this.horizontalMargin = margin }
+
+    /**
+     * The vertical offset from the anchor view to the tooltip when the bounds of the tooltip exceed
+     * the bounds of the screen.
+     *
+     * @param margin - the margin in dips.
+     */
+    fun verticalMarginClipped(margin: Int) = apply { this.verticalMarginClipped = margin }
```

Comment:
Just a curiosity, this apply block is here to allow chaining the setters like in a some sort of builder pattern?

Reply:
Yup, that's exactly it 👍